src/Model/Aggregate.php

```diff
@@ -0,0 +1,299 @@
+ * $aggregate->groupBy(['first','last'], ['salary' => ['sum([])', 'type'=>'money']];
+ *
+ * @property \Atk4\Data\Persistence\Sql $persistence
+ *
+ * @method Expression expr($expr, array $args = []) forwards to Persistence\Sql::expr using $this as model
+ */
+class Aggregate extends Model
+{
+    /** @const string */
+    public const HOOK_INIT_SELECT_QUERY = self::class . '@initSelectQuery';
+
+    /** @var Model */
+    public $baseModel;
+
+    /** @var string[] */
+    public $groupByFields = [];
+
+    /** @var mixed[] */
+    public $aggregateExpressions = [];
+
+    public function __construct(Model $baseModel, array $defaults = [])
+    {
+        if (!$baseModel->persistence instanceof Persistence\Sql) {
+            throw new Exception('Base model must have Sql persistence to use grouping');
+        }
+
+        $this->baseModel = clone $baseModel;
+        $this->table = $baseModel->table;
+
+        // this model does not have ID field
+        $this->id_field = null;
+
+        // this model should always be read-only
+        $this->read_only = true;
+
+        parent::__construct($baseModel->persistence, $defaults);
+
+        // always use table prefixes for this model
+        $this->persistence_data['use_table_prefixes'] = true;
+    }
+
+    /**
+     * Specify a single field or array of fields on which we will group model.
+     *
+     * @param mixed[] $aggregateExpressions Array of aggregate expressions with alias as key
+     *
+     * @return $this
+     */
+    public function groupBy(array $fields, array $aggregateExpressions = [])
+    {
+        $this->groupByFields = array_unique(array_merge($this->groupByFields, $fields));
+
+        foreach ($fields as $fieldName) {
+            $this->addField($fieldName);
+        }
+
+        foreach ($aggregateExpressions as $name => $expr) {
+            $this->aggregateExpressions[$name] = $expr;
+
+            $seed = is_array($expr) ? $expr : [$expr];
+
+            $args = [];
+            // if field originally defined in the parent model, then it can be used as part of expression
+            if ($this->baseModel->hasField($name)) {
+                $args = [$this->baseModel->getField($name)];
+            }
+
+            $seed['expr'] = $this->baseModel->expr($seed[0] ?? $seed['expr'], $args);
+
+            // now add the expressions here
+            $this->addExpression($name, $seed);
+        }
+
+        return $this;
+    }
+
+    public function getRef(string $link): Reference
+    {
+        return $this->baseModel->getRef($link);
+    }
+
+    /**
+     * Method to enable commutative usage of methods enabling both of below
+     * Resulting in Aggregate on $model.
+     *
+     * $model->groupBy(['abc'])->withAggregateField('xyz');
+     *
+     * and
+     *
+     * $model->withAggregateField('xyz')->groupBy(['abc']);
+     */
+    public function withAggregateField(string $name, $seed = []): Model
+    {
+        static::addField($name, $seed);
```

Comment:
I wonder why not `$this->addField`.

Let's finish all feedback to this PR and merge it.

Reply:
method removed